File: LEETCODE/HARD/DAY 11/DAY 16/READ N CHAR GIVEN READ4 - MULTIPLE CALLS.py
# Simulated API
def read4(buf4):
    global file_data, file_ptr
    count = 0
    while count < 4 and file_ptr < len(file_data):
        buf4[count] = file_data[file_ptr]
        file_ptr += 1
        count += 1
    return count

class Reader:
    def __init__(self):
        self.buffer = [""] * 4
        self.offset = 0
        self.bufsize = 0

    def read(self, buf, n):
        readBytes = 0
        eof = False

        while not eof and readBytes < n:
            sz = self.bufsize if self.bufsize > 0 else read4(self.buffer)

            if self.bufsize == 0 and sz < 4:
                eof = True

            bytes_to_copy = min(n - readBytes, sz)
            for i in range(bytes_to_copy):
                buf.append(self.buffer[self.offset + i])

            self.offset = (self.offset + bytes_to_copy) % 4
            self.bufsize = sz - bytes_to_copy
            readBytes += bytes_to_copy

        return readBytes

File: LEETCODE/HARD/DAY 16/READ N CHAR GIVEN READ4 - MULTIPLE CALLS.java
/* The read4 API is defined in the parent class Reader4.
 * int read4(char[] buf);
 */
public class Solution extends Reader4 {
    private char[] buffer = new char[4]; // temp buffer
    private int offset = 0;              // next position in buffer
    private int bufsize = 0;             // number of leftover chars

    /**
     * @param buf Destination buffer
     * @param n   Max number of chars to read
     * @return    Number of chars actually read
     */
    public int read(char[] buf, int n) {
        int readBytes = 0;  // total chars read so far
        boolean eof = false;

        while (!eof && readBytes < n) {
            int sz = (bufsize > 0) ? bufsize : read4(buffer);

            if (bufsize == 0 && sz < 4) {
                eof = true; // reached end of file
            }

            int bytes = Math.min(n - readBytes, sz);
            System.arraycopy(buffer, offset, buf, readBytes, bytes);

            offset = (offset + bytes) % 4;
            bufsize = sz - bytes;
            readBytes += bytes;
        }

        return readBytes;
    }
}
